crates/router/src/services/authentication.rs

```diff
@@ -128,10 +156,32 @@ where
         let admin_api_key =
             get_api_key(request_headers).change_context(errors::ApiErrorResponse::Unauthorized)?;
         let conf = state.conf();
-        if admin_api_key != conf.secrets.admin_api_key {
-            Err(report!(errors::ApiErrorResponse::Unauthorized)
-                .attach_printable("Admin Authentication Failure"))?;
+
+        #[cfg(feature = "kms")]
+        {
+            let hashed_admin_api_key = {
+                get_admin_api_key(
+                    &conf.secrets,
+                    #[cfg(feature = "kms")]
+                    &conf.kms,
+                )
+                .await?
+            };
+
+            if admin_api_key != hashed_admin_api_key.peek() {
+                Err(report!(errors::ApiErrorResponse::Unauthorized)
+                    .attach_printable("Admin Authentication Failure"))?;
+            }
         }
+
+        #[cfg(not(feature = "kms"))]
+        {
+            if admin_api_key != conf.secrets.admin_api_key {
+                Err(report!(errors::ApiErrorResponse::Unauthorized)
+                    .attach_printable("Admin Authentication Failure"))?;
+            }
+        }
```

Comment:
You don't need the feature-specific blocks here, since it's handled in `get_admin_api_key()`.

```suggestion
        let request_admin_api_key =
            get_api_key(request_headers).change_context(errors::ApiErrorResponse::Unauthorized)?;
        let conf = state.conf();

        let admin_api_key = get_admin_api_key(
            &conf.secrets,
            #[cfg(feature = "kms")]
            &conf.kms,
        )
        .await?;

        if request_admin_api_key != admin_api_key.peek() {
            Err(report!(errors::ApiErrorResponse::Unauthorized)
                .attach_printable("Admin Authentication Failure"))?;
        }
```